opsdroid/connector/__init__.py

```diff
@@ -56,60 +103,38 @@ def configuration(self):
         As the method should include some kind of `while True` all messages
         from the chat service should be "awaited" asyncronously to avoid
         blocking the thread.
-
         """
         raise NotImplementedError
 
-    async def respond(self, message, room=None):
-        """Send a message back to the chat service.
-
-        The message object will have a `text` property which should be sent
-        back to the chat service. It may also have a `room` and `user` property
-        which gives information on where the message should be directed.
-
-        Args:
-            message (Message): A message received by the connector.
-            room (string): Name of the room to send the message to
-
-        Returns:
-            bool: True for message successfully sent. False otherwise.
-
+    async def _unknown_event(self, event, target=None):
         """
-        raise NotImplementedError
-
-    async def react(self, message, emoji):
-        """React to a message.
-
-        Not all connectors will have capabilities to react messages, so this
-        method don't have to be implemented and by default logs a debug message
-        and returns False.
-
-        Args:
-            message (Message): A message received by the connector.
-            emoji    (string): The emoji name with which opsdroid will react
-
-        Returns:
-            bool: True for message successfully sent. False otherwise.
-
+        This is the fallback function that is called if the subclass can not
+        handle the event type.
         """
-        _LOGGER.debug(_("%s connector can't react to messages"), self.name)
-        return False
+        raise TypeError(
+            "Connector {type(self)} can not handle the"
+            " '{type(event).__name__}' event type.".format(self=self,
+                                                           event=event))
 
-    async def user_typing(self, trigger):
-        """Signals that opsdroid is typing.
+    async def send(self, event, target=None):
+        """Send a message to the chat service.
 
         Args:
-            opsdroid (OpsDroid): An instance of the opsdroid core.
-            trigger: a bool that allows the event to be triggered on/off
+            event (Event): A message received by the connector.
+            target (string): The name of the room or other place to send the
+            event.
 
-        Triggers the "user is typing" event if the chat service that
-        opsdroid is connected to accepts it.
+        Returns:
+            bool: True for event successfully sent. False otherwise.
         """
+        target = target if target else event.target if event.target else self.default_target
```

Comment:
You're a monster.

Reply:
This is some of the least readable code that I've ever seen 😆. Please don't do things like this, it makes debugging a chore.

Reply:
I agree that it's insane. However, the other options I can think of are pretty much equally insane:

```python
        target_arg = target
        target = self.default_target
        target = event.target if event.target else target
        target = target_arg if target_arg else target
```

```python
        target_arg = target
        target = self.default_target
        if event.target:
            target = event.target
        if target_arg:
            target = target_arg
```

I am open to other suggestions.

Reply:
of course there is always:

```python
list(filter(bool, (target, event.target, self.default_target)))[0]
```

:rofl:

Reply:
How about

```python
target_arg = target
target = target_arg or event.target or self.default_target
```

Reply:
:scream_cat: I did not know you could do that. That's clearly better.

Reply:
When using an `or` in assignment it evaluates the truthiness of each thing being assigned and chooses the first truthy one. So the `None` type variables will just pass to the right.

You can fall into a trap with is because sometimes you want to assign a falsy value. If one of the variables contained an integer `0`, an empty list `[]` or dict `{}` or a boolean `False` it will not assign. But as we know the target will either be an `Event` based object or `None` we are safe.

Reply:
Your mega one line if was so impenetrable I didn't realize this is what you were trying to do, otherwise I would've suggested it sooner 😆.